espnet/nets/pytorch_backend/beam_search.py

```diff
@@ -0,0 +1,297 @@
+
+        Returns:
+            state: pruned decoder state
+        """
+        raise NotImplementedError
+
+    def score(self, y, next_tokens, state, x):
+        """Score new token
+
+        Args:
+            y (torch.Tensor): torch.int64 prefix token (U)
+            next_tokens (torch.Tensor): torch.int64 next token to score (N)
+            state: decoder state for prefix tokens
+            x (torch.Tensor): encoder feature that generates ys (T, D)
+
+        Returns:
+            tuple[torch.Tensor, list[dict]]: Tuple of
+                torch.float32 scores for y (N)
+                and next state for ys
+        """
+        raise NotImplementedError
+
+
+class LengthBonus(DecoderInterface):
+    """Length bonus in beam search"""
+
+    def __init__(self, n_vocab):
+        self.n = n_vocab
+
+    def score(self, y, state, x):
+        """Score new token
+
+        Args:
+            y (torch.Tensor): torch.int64 prefix token to score (B)
+            state: decoder state for prefix tokens
+            x (torch.Tensor): encoder feature that generates ys (T, D)
+
+        Returns:
+            tuple[torch.Tensor, list[dict]]: Tuple of
+                torch.float32 scores for y (B)
+                and next state for ys
+        """
+        import torch
+        return torch.tensor([1.0]).expand(self.n), None
+
+
+def get_weighted_decoders(
+        decoders: Dict[str, DecoderInterface],
+        weights: Dict[str, float]
+) -> Dict[str, Tuple[DecoderInterface, float]]:
+    full_dec_weights = dict()
+    part_dec_weights = dict()
+    """filter invalid weights and decoders"""
+    for k, v in decoders.items():
+        w = weights.get(k, 1.0)
+        if w == 0 or v is None:
+            continue
+        assert isinstance(v, DecoderInterface), f"{k} ({type(v)}) does not implement DecoderInterface"
+        if isinstance(v, PartialDecoderInterface):
+            part_dec_weights[k] = (v, w)
+        else:
+            full_dec_weights[k] = (v, w)
+    return full_dec_weights, part_dec_weights
+
+
+class Hypothesis(NamedTuple):
+    """Hypothesis data type"""
+
+    yseq: torch.Tensor
+    score: float = 0
+    scores: Dict[str, float] = dict()
+    states: Dict[str, Dict] = dict()
+
+    def asdict(self):
+        return self._replace(yseq=self.yseq.tolist())._asdict()
+
+
+def append_tensor(xs, x):
+    x = torch.tensor([x], dtype=xs.dtype, device=xs.device)
+    return torch.cat((xs, x))
+
+
+def beam_search(x, sos, eos, beam_size, decoders, weights,
+                token_list=None, maxlenratio=0.0, minlenratio=0.0,
+                pre_beam_ratio=1.5, pre_beam_score="decoder"):
+    """Beam search with scorers
+
+    Args:
+        x (torch.Tensor): Encoded speech feature (T, D)
+        sos (int): Start of sequence id
+        eos (int): End of sequence id
+        beam_size (int): The number of hypotheses kept during search
+        decoders (dict[str, DecoderInterface]): list of decoder modules
```

Comment:
Can you add an example of multiple decoders here (attention encoder-decoder, CTC, RNNLM?)?